File: problems/valid_palindrome/valid_palindrome_test.py
"""Test cases for the isPalindrome function."""

import pytest

from valid_palindrome import isPalindrome


@pytest.mark.parametrize(
    "s, expected",
    [
        (
            "A man, a plan, a canal: Panama",
            True,
        ),  # Classic palindrome with spaces and punctuation
        ("race a car", False),  # Not a palindrome
        ("", True),  # Empty string
        ("a", True),  # Single character
        ("Aa", True),  # Case insensitive palindrome
        ("ab", False),  # Simple two character non-palindrome
        ("aa", True),  # Simple two character palindrome
        ("racecar", True),  # Simple palindrome
        ("hello", False),  # Simple non-palindrome
        ("A Santa at NASA", True),  # Palindrome with spaces
        ("Was it a car or a cat I saw?", True),  # Palindrome with punctuation
        ("No 'x' in Nixon", True),  # Palindrome with apostrophe
        ("Madam, I'm Adam", True),  # Palindrome with comma and apostrophe
        ("never odd or even", True),  # Palindrome with spaces
        ("nope", False),  # Simple non-palindrome
        ("0P", False),  # Mixed alphanumeric non-palindrome
        ("A man, a plan, a canal: Panama!", True),  # With exclamation
        ("race a E-car", True),  # With hyphen and mixed case
        ("Able was I ere I saw Elba", True),  # Historical palindrome
        ("12321", True),  # Numeric palindrome
        ("12345", False),  # Numeric non-palindrome
        ("a1b2c3c2b1a", True),  # Alphanumeric palindrome
        ("1a2b3c3c2b1a", False),  # Alphanumeric non-palindrome
        (".,!@#$%^&*()", True),  # Only special characters (empty after cleaning)
        ("a.,!@#$%^&*()a", True),  # Palindrome with special chars in middle
        ("Dammit, I'm mad!", True),  # Another classic palindrome
        ("Step on no pets", True),  # Word-based palindrome
        ("Was it a rat I saw?", True),  # Question mark palindrome
        ("Mr. Owl ate my metal worm", True),  # Complex palindrome
    ],
    ids=[
        "classic_panama",
        "race_car_false",
        "empty_string",
        "single_char",
        "case_insensitive",
        "two_char_false",
        "two_char_true",
        "simple_palindrome",
        "simple_false",
        "santa_nasa",
        "car_cat_question",
        "nixon_apostrophe",
        "madam_adam",
        "never_odd_even",
        "nope_false",
        "mixed_alphanum_false",
        "panama_exclamation",
        "race_ecar_hyphen",
        "able_elba",
        "numeric_palindrome",
        "numeric_false",
        "alphanum_palindrome",
        "alphanum_false",
        "only_special_chars",
        "special_chars_middle",
        "dammit_mad",
        "step_pets",
        "rat_question",
        "owl_worm",
    ],
)
def test_is_palindrome(s, expected):
    assert isPalindrome(s) == expected

File: problems/valid_palindrome/valid_palindrome_test.cc
#include "valid_palindrome.h"

#include <gtest/gtest.h>
#include <string>

using std::string;

struct ValidPalindromeCase {
  const std::string test_name;
  const string s;
  const bool expected;
};

using ValidPalindromeTest = ::testing::TestWithParam<ValidPalindromeCase>;

TEST_P(ValidPalindromeTest, TestCases) {
  const ValidPalindromeCase &test_case = GetParam();
  const bool result = isPalindrome(test_case.s);
  EXPECT_EQ(result, test_case.expected);
}

INSTANTIATE_TEST_SUITE_P(
    ValidPalindromeTestCases, ValidPalindromeTest,
    ::testing::Values(
        ValidPalindromeCase{.test_name = "ClassicPanama",
                            .s = "A man, a plan, a canal: Panama",
                            .expected = true},
        ValidPalindromeCase{
            .test_name = "RaceCarFalse", .s = "race a car", .expected = false},
        ValidPalindromeCase{
            .test_name = "EmptyString", .s = "", .expected = true},
        ValidPalindromeCase{
            .test_name = "SingleChar", .s = "a", .expected = true},
        ValidPalindromeCase{
            .test_name = "CaseInsensitive", .s = "Aa", .expected = true},
        ValidPalindromeCase{
            .test_name = "TwoCharFalse", .s = "ab", .expected = false},
        ValidPalindromeCase{
            .test_name = "TwoCharTrue", .s = "aa", .expected = true},
        ValidPalindromeCase{
            .test_name = "SimplePalindrome", .s = "racecar", .expected = true},
        ValidPalindromeCase{
            .test_name = "SimpleFalse", .s = "hello", .expected = false},
        ValidPalindromeCase{
            .test_name = "SantaNasa", .s = "A Santa at NASA", .expected = true},
        ValidPalindromeCase{.test_name = "CarCatQuestion",
                            .s = "Was it a car or a cat I saw?",
                            .expected = true},
        ValidPalindromeCase{.test_name = "NixonApostrophe",
                            .s = "No 'x' in Nixon",
                            .expected = true},
        ValidPalindromeCase{
            .test_name = "MadamAdam", .s = "Madam, I'm Adam", .expected = true},
        ValidPalindromeCase{.test_name = "NeverOddEven",
                            .s = "never odd or even",
                            .expected = true},
        ValidPalindromeCase{
            .test_name = "NopeFalse", .s = "nope", .expected = false},
        ValidPalindromeCase{
            .test_name = "MixedAlphanumFalse", .s = "0P", .expected = false},
        ValidPalindromeCase{.test_name = "PanamaExclamation",
                            .s = "A man, a plan, a canal: Panama!",
                            .expected = true},
        ValidPalindromeCase{.test_name = "RaceEcarHyphen",
                            .s = "race a E-car",
                            .expected = true},
        ValidPalindromeCase{.test_name = "AbleElba",
                            .s = "Able was I ere I saw Elba",
                            .expected = true},
        ValidPalindromeCase{
            .test_name = "NumericPalindrome", .s = "12321", .expected = true},
        ValidPalindromeCase{
            .test_name = "NumericFalse", .s = "12345", .expected = false},
        ValidPalindromeCase{.test_name = "AlphanumPalindrome",
                            .s = "a1b2c3c2b1a",
                            .expected = true},
        ValidPalindromeCase{.test_name = "AlphanumFalse",
                            .s = "1a2b3c3c2b1a",
                            .expected = false},
        ValidPalindromeCase{.test_name = "OnlySpecialChars",
                            .s = ".,!@#$%^&*()",
                            .expected = true},
        ValidPalindromeCase{.test_name = "SpecialCharsMiddle",
                            .s = "a.,!@#$%^&*()a",
                            .expected = true},
        ValidPalindromeCase{.test_name = "DammitMad",
                            .s = "Dammit, I'm mad!",
                            .expected = true},
        ValidPalindromeCase{
            .test_name = "StepPets", .s = "Step on no pets", .expected = true},
        ValidPalindromeCase{.test_name = "RatQuestion",
                            .s = "Was it a rat I saw?",
                            .expected = true},
        ValidPalindromeCase{.test_name = "OwlWorm",
                            .s = "Mr. Owl ate my metal worm",
                            .expected = true}),
    [](const ::testing::TestParamInfo<ValidPalindromeCase> &info) {
      return info.param.test_name;
    });